File: academic/cpp/src/fila_requisicoes.cpp
#include "fila_requisicoes.hpp"

#include <stdexcept>

static const char* PRIORIDADES[QUANTIDADE_DE_PRIORIDADES] = {"EMERGENCIA", "URGENTE", "ELETIVA"};

FilaRequisicoes::FilaRequisicoes() : quantidade(0) {
    for (int faixa = 0; faixa < QUANTIDADE_DE_PRIORIDADES; faixa++) {
        cabecas[faixa] = nullptr;
        caudas[faixa] = nullptr;
    }
}

FilaRequisicoes::~FilaRequisicoes() {
    limpar();
}

int FilaRequisicoes::indiceDaPrioridade(const std::string& prioridade) {
    for (int faixa = 0; faixa < QUANTIDADE_DE_PRIORIDADES; faixa++) {
        if (prioridade == PRIORIDADES[faixa]) {
            return faixa;
        }
    }
    throw std::invalid_argument("Prioridade desconhecida na fila: " + prioridade);
}

void FilaRequisicoes::enfileirar(const RequisicaoAed& requisicao) {
    const int faixa = indiceDaPrioridade(requisicao.prioridade);
    NoRequisicao* novo = new NoRequisicao(requisicao);
    if (cabecas[faixa] == nullptr) {
        cabecas[faixa] = novo;
        caudas[faixa] = novo;
    } else {
        caudas[faixa]->proximo = novo;
        caudas[faixa] = novo;
    }
    quantidade++;
}

bool FilaRequisicoes::desenfileirar(RequisicaoAed& retirada) {
    for (int faixa = 0; faixa < QUANTIDADE_DE_PRIORIDADES; faixa++) {
        NoRequisicao* cabeca = cabecas[faixa];
        if (cabeca != nullptr) {
            cabecas[faixa] = cabeca->proximo;
            if (cabecas[faixa] == nullptr) {
                caudas[faixa] = nullptr;
            }
            retirada = cabeca->dado;
            delete cabeca;
            quantidade--;
            return true;
        }
    }
    return false;
}

bool FilaRequisicoes::espiar(RequisicaoAed& proxima) const {
    for (int faixa = 0; faixa < QUANTIDADE_DE_PRIORIDADES; faixa++) {
        if (cabecas[faixa] != nullptr) {
            proxima = cabecas[faixa]->dado;
            return true;
        }
    }
    return false;
}

int FilaRequisicoes::tamanho() const {
    return quantidade;
}

bool FilaRequisicoes::estaVazia() const {
    return quantidade == 0;
}

void FilaRequisicoes::limpar() {
    for (int faixa = 0; faixa < QUANTIDADE_DE_PRIORIDADES; faixa++) {
        NoRequisicao* atual = cabecas[faixa];
        while (atual != nullptr) {
            NoRequisicao* proximo = atual->proximo;
            delete atual;
            atual = proximo;
        }
        cabecas[faixa] = nullptr;
        caudas[faixa] = nullptr;
    }
    quantidade = 0;
}

File: src/main/java/br/edu/cesar/bloodell/aed/u1/FilaRequisicoes.java
package br.edu.cesar.bloodell.aed.u1;

import br.edu.cesar.bloodell.aed.contratos.RequisicaoAed;

public class FilaRequisicoes {

    private static final String[] PRIORIDADES = {"EMERGENCIA", "URGENTE", "ELETIVA"};

    private final NoRequisicao[] cabecas = new NoRequisicao[PRIORIDADES.length];
    private final NoRequisicao[] caudas = new NoRequisicao[PRIORIDADES.length];
    private int quantidade;

    public FilaRequisicoes() {
        this.quantidade = 0;
    }

    public void enfileirar(RequisicaoAed requisicao) {
        if (requisicao == null) {
            throw new IllegalArgumentException("Nao e possivel enfileirar requisicao nula.");
        }
        int faixa = indiceDaPrioridade(requisicao.getPrioridade());
        NoRequisicao novo = new NoRequisicao(requisicao);
        if (cabecas[faixa] == null) {
            cabecas[faixa] = novo;
            caudas[faixa] = novo;
        } else {
            caudas[faixa].setProximo(novo);
            caudas[faixa] = novo;
        }
        quantidade++;
    }

    public RequisicaoAed desenfileirar() {
        for (int faixa = 0; faixa < PRIORIDADES.length; faixa++) {
            NoRequisicao cabeca = cabecas[faixa];
            if (cabeca != null) {
                cabecas[faixa] = cabeca.getProximo();
                if (cabecas[faixa] == null) {
                    caudas[faixa] = null;
                }
                cabeca.setProximo(null);
                quantidade--;
                return cabeca.getDado();
            }
        }
        return null;
    }

    public RequisicaoAed espiar() {
        for (int faixa = 0; faixa < PRIORIDADES.length; faixa++) {
            if (cabecas[faixa] != null) {
                return cabecas[faixa].getDado();
            }
        }
        return null;
    }

    public RequisicaoAed[] paraVetor() {
        RequisicaoAed[] itens = new RequisicaoAed[quantidade];
        int indice = 0;
        for (int faixa = 0; faixa < PRIORIDADES.length; faixa++) {
            NoRequisicao atual = cabecas[faixa];
            while (atual != null) {
                itens[indice] = atual.getDado();
                indice++;
                atual = atual.getProximo();
            }
        }
        return itens;
    }

    public int tamanho() {
        return quantidade;
    }

    public boolean estaVazia() {
        return quantidade == 0;
    }

    public void limpar() {
        for (int faixa = 0; faixa < PRIORIDADES.length; faixa++) {
            cabecas[faixa] = null;
            caudas[faixa] = null;
        }
        quantidade = 0;
    }

    private int indiceDaPrioridade(String prioridade) {
        for (int faixa = 0; faixa < PRIORIDADES.length; faixa++) {
            if (PRIORIDADES[faixa].equals(prioridade)) {
                return faixa;
            }
        }
        throw new IllegalArgumentException("Prioridade desconhecida na fila: " + prioridade);
    }
}